src/security/KeyData.cc

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 1996-2017 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "anyp/PortCfg.h"
+#include "fatal.h"
+#include "security/KeyData.h"
+#include "SquidConfig.h"
+#include "ssl/bio.h"
+
+/// verify that a private key and cert match
+static bool
+checkPrivateKey(Security::CertPointer &cert, Security::PrivateKeyPointer &pkey)
+{
+#if USE_OPENSSL
+    return X509_check_private_key(cert.get(), pkey.get());
+#elif USE_GNUTLS
+    return true; // TODO find GnuTLS equivalent check
+#else
+    return false;
+#endif
+}
+
+/**
+ * Read certificate from file.
+ * See also: Ssl::ReadX509Certificate function, gadgets.cc file
+ */
+bool
+Security::KeyData::loadX509CertFromFile()
+{
+#if USE_OPENSSL
+    const char *certFilename = certFile.c_str();
+    Ssl::BIO_Pointer bio(BIO_new(BIO_s_file()));
+    if (!bio || !BIO_read_filename(bio.get(), certFilename)) {
+        const auto x = ERR_get_error();
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to load file '" << certFile << "': " << ErrorString(x));
+        return false;
+    }
+
+    X509 *certificate = PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr);
+
+#elif USE_GNUTLS
+    const char *certFilename = certFile.c_str();
+    gnutls_datum_t data;
+    Security::ErrorCode x = gnutls_load_file(certFilename, &data);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to load file '" << certFile << "': " << ErrorString(x));
+        return false;
+    }
+
+    gnutls_pcert_st pcrt;
+    x = gnutls_pcert_import_x509_raw(&pcrt, &data, GNUTLS_X509_FMT_PEM, 0);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to import certificate from '" << certFile << "': " << ErrorString(x));
+        return false;
+    }
+    gnutls_free(data.data);
+
+    gnutls_x509_crt_t certificate;
+    x = gnutls_pcert_export_x509(&pcrt, &certificate);
+    if (x != GNUTLS_E_SUCCESS) {
+        certificate = nullptr; // paranoid: just in case the *_t ptr is undefined after deinit.
```

Comment:
AFAIK this function call is just a data copy from one structure to another. So it fails only if the input was loaded with non-X509 types (ie. incompatible with X.509). We are loading the data immediately prior with explicit GNUTLS_X509_FMT_PEM flag to ensure that only X.509 succeeds in the earlier call.

The remaining edge case is a non-existing certificate. The error for that is in the else clause of the next if-statement block shared by both GnuTLS and OpenSSL, all we need do to trigger that is set the pointer to nil (as done for the paranoid case)

Reply:
I am surprised that you are trying to justify the lack of error reporting by speculating that the function usually does not fail. Looking at GnuTLS sources, I quickly lost count of the different error conditions gnutls_pcert_export_x509() and its callees detect! The function is a lot more complex than a simple copy and even if it were that simple, the simplicity is not a valid excuse to avoid error reporting, especially in code that already detects the error condition and that already reports other similar errors.

Reply:
fine then, added.